### PR TITLE
Export findings to CSV and Excel

### PR DESCRIPTION
**Pull request by StefanFl**
The finding lists now provide an export functionality to download CSV and Excel files with the filtered findings:
2021-09-24 08_37_45-Findings _ DefectDojo

This is a solution for #4900

Sample reports:
- [findings.xlsx](https://github.com/DefectDojo/django-DefectDojo/files/7224349/findings.xlsx)
- [findings.csv](https://github.com/DefectDojo/django-DefectDojo/files/7224350/findings.csv)

This PR uses the implementation to select findings that was introduced with the quick reports. This implementation didn't respect the filters and reported too many findings. This has been fixed with one exception: Filtering for tags and finding_groups is not trivial and is not yet implemented.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -73,3 +73,4 @@ django-test-migrations==1.1.0
 djangosaml2==1.3.3
 drf-spectacular==0.19.0
 django-ratelimit==3.0.1
+XlsxWriter==3.0.1
```
Comment (damiencarol) [inline requirements.txt:76]: We have already have a dependency on module `openpyxl`.

**Review by damiencarol — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 1 above)
Comment (StefanFl) [reply requirements.txt:76]: Yes, we should have only one library. I will have a look which one to keep.

**Comment by valentijnscholten** [comment]
Is there a reason for the custom url / query parsing (introduced earlier). I would say `django-filters` should do that for us?

**Review by valentijnscholten — CHANGES_REQUESTED** [review]
Can you add some integration tests that trigger the exports?

**Comment by StefanFl** [comment]
> Is there a reason for the custom url / query parsing (introduced earlier). I would say django-filters should do that for us?

I used it because it's there. I don't know how to achieve filtering this part of the query string with django-filter. Someone else would have to take over to change that, but maybe in a second step when this PR is merged.

**Comment by devGregA** [comment]
Does the current CSV export not capture all the findings (bottom left of screenshot)?

**Comment by StefanFl** [comment]
> Does the current CSV export not capture all the findings (bottom left of screenshot)?

These exports only export the findings that are currently on the screen with the columns in the list. If we wanted to have more attributes exported with that, we would have to put all the attributes in the HTML list. 

I don't know where these buttons come from.

**Comment by devGregA** [comment]
Interesting, just played around and tested the old buttons. Stefan's version seems much better. Not asking Stefan too, but we should likely remove those in the future, as I'm not sure what the value even is of grabbing things from a single page.

File: requirements.txt (hunk 1 above)
Comment (StefanFl) [reply requirements.txt]: The Excel export is now using openpyxl and I removed XlsxWriter from the requirements.

**Comment by StefanFl** [comment]
> Can you add some integration tests that trigger the exports?

Done

**Review by devGregA — APPROVED** [review]

**Comment by valentijnscholten** [comment]
The pre-existing records come from the javascript datatables plugin. I guess someone added those thinking "something is better then nothing". It doesn't really hurt, but can be a bit confusing at first. The PDF button can be helpful if you have a small amount of findings and you want to do a quick report. It looks ugly though as it has no stylesheet.

**Review by valentijnscholten — DISMISSED** [review]

**Comment by StefanFl** [comment]
> Interesting, just played around and tested the old buttons. Stefan's version seems much better. Not asking Stefan too, but we should likely remove those in the future, as I'm not sure what the value even is of grabbing things from a single page.

This is actually really easy to do, I found the buttons now and removed them with my latest commit to this PR.

File: tests/finding_test.py [hunk 2]
```diff
@@ -362,9 +394,13 @@ def add_finding_tests_to_suite(suite, jira=False, github=False, block_execution=
     suite.addTest(ProductTest('test_add_product_finding'))
     # TODO add some more findings with different statuses
     suite.addTest(FindingTest('test_list_findings_all'))
-    suite.addTest(FindingTest('test_list_findings_closed'))
-    suite.addTest(FindingTest('test_list_findings_accepted'))
-    suite.addTest(FindingTest('test_list_findings_open'))
+    # The next 3 tests don't work and need to be reworked
+    # suite.addTest(FindingTest('test_list_findings_closed'))
+    # suite.addTest(FindingTest('test_list_findings_accepted'))
+    # suite.addTest(FindingTest('test_list_findings_open'))
```
Comment (valentijnscholten) [inline tests/finding_test.py:400]: what's wrong with the tests, I can help maybe if needed to fix them.

File: tests/finding_test.py [hunk 3]
```diff
@@ -51,6 +53,36 @@ def test_list_findings(self, suffix):
...
+
+    def test_excel_export(self):
+        driver = self.driver
+        driver.get(self.base_url + "finding")
+
+        driver.find_element_by_id("downloadMenu").click()
+        driver.find_element_by_id("excel_export").click()
+
+        time.sleep(2)
+        self.assertTrue(Path("/tmp/findings.xlsx").is_file())
```
Comment (valentijnscholten) [inline tests/finding_test.py:84]: I think this is a server side assert? And I guess you looked at doing a client side quick, but it's not possible or hard?

File: tests/base_test_class.py [hunk 4]
```diff
@@ -321,7 +325,7 @@ def assertNoConsoleErrors(self):
 
             The addition of the trigger exception is due to the Report Builder tests.
             """
-            accepted_javascript_messages = r'(zoom\-in\.cur.*)404\ \(Not\ Found\)|Uncaught TypeError: Cannot read properties of null \(reading \'trigger\'\)'
+            accepted_javascript_messages = r'(zoom\-in\.cur.*)404\ \(Not\ Found\)|Uncaught TypeError: Cannot read properties of null \(reading \'trigger\'\)|Uncaught TypeError: Cannot read properties of null \(reading \'innerHTML\'\)'
```
Comment (valentijnscholten) [inline tests/base_test_class.py:328]: should we add a comment to explain where this new "ignore exception" is coming from?

File: tests/finding_test.py (hunk 3 above)
Comment (valentijnscholten) [reply tests/finding_test.py:84]: or maybe it is client side, just saw the prefs now above

**Comment by valentijnscholten** [comment]
Great PR, but still some comments from me :-)

File: tests/finding_test.py (hunk 2 above)
Comment (StefanFl) [reply tests/finding_test.py:400]: I didn't look what's wrong with them. They got never executed properly. Each of these tests calls `test_list_findings` with a parameter for the suffix of the URL like `finding/open` or `finding/closed`. But in `test_list_findings` this suffix is ignored and it executes the same test for all findings 4 times. Have a look at https://github.com/DefectDojo/django-DefectDojo/blob/master/tests/finding_test.py#L17-L32 and you will see.

It's completely unrelated with this PR, just something I noted when I saw it while writing more tests.

Comment (StefanFl) [reply tests/finding_test.py:400]: It should be fixed in a separate PR

Comment (valentijnscholten) [reply tests/finding_test.py:400]: ah oke, at first flance it looked like this PR broke. maybe put that comment in there and leave it for another PR indeed.

File: tests/finding_test.py (hunk 3 above)
Comment (StefanFl) [reply tests/finding_test.py:84]: What do you mean with client or server side assert, is there a difference? The implementation comes from the Python `unittest` package and raises an exception when the expression is false:

```
    def assertTrue(self, expr, msg=None):
        """Check that the expression is true."""
        if not expr:
            msg = self._formatMessage(msg, "%s is not true" % safe_repr(expr))
            raise self.failureException(msg)
```

File: tests/base_test_class.py (hunk 4 above)
Comment (StefanFl) [reply tests/base_test_class.py:329]: Yes we should 😃, it's done with https://github.com/DefectDojo/django-DefectDojo/pull/5148/commits/01e854a9105e4d679911cc06aaa0cf2898108256

**Comment by StefanFl** [comment]
Another idea, we could add a list of endpoints, separated by newlines for Excel and semicolon for CSV. Maybe not all of them but max 5 followed by `...` when there are more so this cell doesn't get too long. What do you think?

File: tests/finding_test.py (hunk 3 above)
Comment (valentijnscholten) [reply tests/finding_test.py:84]: In some unit tests there is sometimes a mix of things being checked. In the api tests for example some asserts are on the data returned to the client (so "client-side") and some asserts are on server-side data like things in the database.
At first glance the `/tmp/findings.xlsx` assert looked like "server-side" to me. But I was wrong, so forget about it :-)

File: tests/finding_test.py (hunk 2 above)
Comment (StefanFl) [reply tests/finding_test.py:400]: I wrote issue #5154 to address this.

**Review by damiencarol — APPROVED** [review]

**Comment by StefanFl** [comment]
>>Is there a reason for the custom url / query parsing (introduced earlier). I would say django-filters should do that for us?

> I used it because it's there. I don't know how to achieve filtering this part of the query string with django-filter. Someone else would have to take over to change that, but maybe in a second step when this PR is merged.

see issue  #5155

**Pull request merged**